DEPR: Series/DataFrame.append (#35407)

neinkeinkaffee opened the pull request:
- [x] closes #35407 (adds deprecation warning)
- [x] closes #22957 
- [x] tests added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit) for how to run them
- [x] whatsnew entry

neinkeinkaffee committed:
DEPR: Series/DataFrame.append (#35407)

jreback commented on doc/source/whatsnew/v1.4.0.rst:
```diff
@@ -438,6 +438,61 @@ when given numeric data, but in the future, a :class:`NumericIndex` will be retu
+
+.. code-block:: ipython
+
+    In [1]: pd.core.reshape.concat.concat([pd.Series([1, 2]), pd.Series([3, 4])])
```
use the idiomatic pd.concat

jreback commented on pandas/core/frame.py:
```diff
@@ -9114,6 +9114,14 @@ def append(
+            "and will be removed from pandas in a future version. "
+            "Use pandas.core.reshape.concat instead.",
+            FutureWarning,
+            stacklevel=2,
```
use find_stack_level

jreback commented on pandas/core/series.py:
```diff
@@ -2888,6 +2888,14 @@ def append(
+            "and will be removed from pandas in a future version. "
+            "Use pandas.core.reshape.concat instead.",
+            FutureWarning,
+            stacklevel=2,
```
same

jreback commented on pandas/tests/frame/methods/test_append.py:
```diff
@@ -21,19 +21,22 @@ def test_append_multiindex(self, multiindex_dataframe_random_data, frame_or_seri
         b = obj[5:]
 
-        result = a.append(b)
+        with tm.assert_produces_warning(FutureWarning):
```
would be ok just to filter warnings for this entire class

jreback reviewed (changes requested):
ok to filter warnings and have a single test to assert the deprecation warning

neinkeinkaffee committed:
Refer to pandas.concat instead of pandas.core.reshape.concat

neinkeinkaffee committed:
Add explicit test for warning, ignore in the remainder of the tests

neinkeinkaffee committed:
Refer to concat in root namespace

neinkeinkaffee commented:
Thanks for the suggestions. I've tried to address them. Could you re-review and let me know your feedback? Cheers!

phofl commented on pandas/core/series.py:
```diff
@@ -2888,6 +2888,14 @@ def append(
         ValueError: Indexes have overlapping values: [0, 1, 2]
         """
+        warnings.warn(
+            "The frame.append method is deprecated "
```
Series here

phofl commented on pandas/tests/frame/methods/test_append.py:
```diff
@@ -270,3 +270,9 @@ def test_multiindex_column_append_multiple(self):
             )
             tm.assert_frame_equal(result, expected)
+
+    def test_append_raises_future_warning(self):
```
add issue number

phofl commented on pandas/tests/series/methods/test_append.py:
```diff
@@ -82,7 +83,12 @@ def test_append_dataframe_raises(self):
             df.A.append([df])
 
+    def test_append_raises_future_warning(self):
+        with tm.assert_produces_warning(FutureWarning):
```
issue number

neinkeinkaffee replied:
Thanks! Where does the issue number go? Is there another example for this in the existing tests?

MarcoGorelli replied:
Just below the definition, e.g.
```
def test_append_raises_future_warning(self):
    # GH#44539
```

phofl commented on doc/source/whatsnew/v1.4.0.rst:
```diff
@@ -438,6 +438,61 @@ when given numeric data, but in the future, a :class:`NumericIndex` will be retu
+
+    In [1]: pd.Series([1, 2]).append(pd.Series([3, 4])
+    Out [1]:
+    <stdin>:1: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
```
series.append

neinkeinkaffee committed:
Fix copy-paste errors

phofl reviewed (changes requested):
Web and docs is failing because of the warnings

neinkeinkaffee committed:
Ignore warning in reshape/concat test_append and add issue reference

neinkeinkaffee committed:
Ignore warnings in reshape test_crosstab

neinkeinkaffee commented:
Question on the failing web and docs: looking at the failing doc build, some of the cases are due to direct calls to append, which I can change into concat. But others are indirect, e.g. `doc/source/info_tutorials/getting_started/02_read_write.rst` calls titanic.info() on a DataFrame, which calls frame.memory_usage which calls series.append. In such cases, it's probably preferred to change the series.append into a concat in frame.memory_usage? Or should I just add an `:okwarning:` to the ipython codeblock?

jreback commented:
no all usages should be changes in tests and docs except those we r explicitly doing

neinkeinkaffee commented:
@jreback Just to make sure I don't misunderstand: given that frame.memory_usage is neither in doc nor test, I shouldn't change the call to series.append to concat there, but instead add a :okwarning: to the ipython codeblock? And likewise not change any other instances were append is called in the pandas code that's neither documentation nor test?

jreback commented:
no u do not add okwarning anywhere 

u need to fix this for example in the code not to output a warning (as this is tested in many tests)

jreback commented:
when you push pls merge master

neinkeinkaffee commented:
Yes, will do!

neinkeinkaffee committed:
Replace further appends with concat

neinkeinkaffee committed:
Merge branch 'master' into 35407-deprecate-append

neinkeinkaffee committed:
Ignore FutureWarning in test_value_counts_null

neinkeinkaffee committed:
Merge branch 'master' into 35407-deprecate-append

neinkeinkaffee committed:
Filter FutureWarning in remaining affected tests

neinkeinkaffee committed:
Ignore FutureWarning in even more tests

neinkeinkaffee committed:
Ignore FutureWarning in one last test

neinkeinkaffee committed:
Delete refs to merging.concatenation and merging.append.row (documenting append)

jreback commented on doc/source/whatsnew/v1.4.0.rst:
```diff
@@ -490,6 +490,61 @@ when given numeric data, but in the future, a :class:`NumericIndex` will be retu
+
+*Recommended syntax*
+
+.. code-block:: ipython
```
dont' use a code-block here, use an ipython-block (so it runs the code)

jreback commented on pandas/core/indexing.py:
```diff
@@ -1993,7 +1993,14 @@ def _setitem_with_indexer_missing(self, indexer, value):
-                self.obj._mgr = self.obj.append(value)._mgr
+                from pandas.core.reshape.concat import concat
+
+                df = value.to_frame().T
```
hmm this not great. cc @jbrockmendel can you suggest anything better here

jbrockmendel replied:
this looks very nearly cut/paste from DataFrame.append.  the only difference is that we actually have _tests_ for DataFrame.append.  If this code has to exist somewhere, then we might as well just implement DataFrame._append so we can adapt the existing tests (and im guessing the scope of DataFrame._append would be narrower so once the deprecation is enforced we could rip out a bunch of it)

jreback commented on pandas/tests/base/test_value_counts.py:
```diff
@@ -38,6 +38,7 @@ def test_value_counts(index_or_series_obj):
     tm.assert_series_equal(result, expected)
 
 
+@pytest.mark.filterwarnings("ignore:.*append method is deprecated.*:FutureWarning")
```
why is this showing a warning? generally *don't* add filterwarnings in anything but the actual method testing itself (e.g. append), these can be suppressing a warning that is thrown internally which we want to fix

jreback commented on pandas/tests/frame/methods/test_append.py:
```diff
@@ -12,6 +12,7 @@
 import pandas._testing as tm
 
 
+@pytest.mark.filterwarnings("ignore:.*append method is deprecated.*:FutureWarning")
```
this is ok!

jreback commented on pandas/tests/frame/methods/test_drop_duplicates.py:
```diff
@@ -21,6 +21,7 @@ def test_drop_duplicates_with_misspelled_column_name(subset):
         df.drop_duplicates(subset)
 
 
+@pytest.mark.filterwarnings("ignore:.*append method is deprecated.*:FutureWarning")
```
not here

jreback commented on pandas/tests/generic/test_duplicate_labels.py:
```diff
@@ -294,6 +294,7 @@ def test_setting_allows_duplicate_labels_raises(self, data):
 
         assert data.flags.allows_duplicate_labels is True
 
+    @pytest.mark.filterwarnings("ignore:.*append method is deprecated.*:FutureWarning")
```
not here!

jreback replied:
remove this

jreback commented on pandas/tests/generic/test_finalize.py:
```diff
@@ -511,6 +511,7 @@ def ndframe_method(request):
     return request.param
 
 
+@pytest.mark.filterwarnings("ignore:.*append method is deprecated.*:FutureWarning")
```
not here! (if append is in this list then ok)

jreback commented on pandas/tests/indexes/period/test_indexing.py:
```diff
@@ -115,6 +115,7 @@ def test_getitem_index(self):
         )
         tm.assert_index_equal(result, exp)
 
+    @pytest.mark.filterwarnings("ignore:.*append method is deprecated.*:FutureWarning")
```
not here!

jreback reviewed (changes requested):
you don't simply want to suppress the warnings anywhere, you need to actually fix code that is causing the warnings.

jbrockmendel commented on pandas/core/frame.py:
```diff
@@ -3265,9 +3265,16 @@ def memory_usage(self, index: bool = True, deep: bool = False) -> Series:
+
+            result = concat(
+                [
+                    self._constructor_sliced(
```
nitpick: can you do `whatever = self._constructor_sliced(...)` in a separate line

jbrockmendel commented on pandas/core/frame.py:
```diff
@@ -9709,7 +9724,11 @@ def c(x):
+                from pandas.core.reshape.concat import concat
+
+                correl = concat(
+                    [correl, Series([np.nan] * len(idx_diff), index=idx_diff)]
```
can you define the extra row on its own line (ideally with an informative name)

neinkeinkaffee committed:
Replace append by concat in tests instead of ignoring warnings

neinkeinkaffee committed:
Introduce intermediate variables

neinkeinkaffee committed:
Use ipython instead of code block

neinkeinkaffee committed:
Merge branch 'master' into 35407-deprecate-append

MarcoGorelli commented:
I think @jbrockmendel 's suggestion

> this looks very nearly cut/paste from DataFrame.append. the only difference is that we actually have tests for DataFrame.append. If this code has to exist somewhere, then we might as well just implement DataFrame._append so we can adapt the existing tests (and im guessing the scope of DataFrame._append would be narrower so once the deprecation is enforced we could rip out a bunch of it)

makes a lot of sense.

So, this would mean:
- rename `append` to `_append`
- inside `append`, just raise a future warning and then call `_append`
- when `append` is used in internal code, replace it with `_append`
- when `append` is used in user-facing code (e.g. docs), replace it with `concat`
?

@neinkeinkaffee apologies, this has ended up being a lot more work than I'd anticipated when I suggested the issue. However, hopefully you're finding it to be worthwhile!

jreback commented:
+1 on @jbrockmendel suggestion

neinkeinkaffee commented:
I find it absolutely worthwhile, thanks @MarcoGorelli. 

Thanks everyone for your comments and patience, really appreciated!

neinkeinkaffee committed:
Extract _append and replace append by _append/concat in tests with ignored warning

neinkeinkaffee committed:
Reinsert modified instructions for appending one row to df (suggestion)

neinkeinkaffee committed:
Merge master

neinkeinkaffee committed:
Import concat from pandas in tests

MarcoGorelli commented on pandas/core/frame.py:
```diff
@@ -3267,9 +3267,17 @@ def memory_usage(self, index: bool = True, deep: bool = False) -> Series:
                 self.index.memory_usage(deep=deep), index=["Index"]
-            ).append(result)
+            )
+            result = concat(
+                [
+                    index_memory_usage,
+                    result,
+                ]
+            )
```
really minor, but can this fit on a single line?

neinkeinkaffee committed:
Fit call to concat on one line

neinkeinkaffee committed:
Merge branch 'master' into 35407-deprecate-append

neinkeinkaffee committed:
Replace append by concat in _add_margins

jreback commented on pandas/tests/indexing/test_partial.py:
```diff
@@ -361,7 +361,7 @@ def test_partial_setting_mixed_dtype(self):
         s = df.loc[1].copy()
         s.name = 2
-        expected = df.append(s)
+        expected = df._append(s)
```
can you try not to use _append in tests and pref rewriting as concat

jreback commented on pandas/tests/reshape/concat/test_categorical.py:
```diff
@@ -184,6 +185,7 @@ def test_concat_categorical_unchanged(self):
         )
         tm.assert_equal(result, expected)
 
+    @pytest.mark.filterwarnings("ignore:.*append method is deprecated.*:FutureWarning")
```
remove

jreback commented on pandas/tests/reshape/concat/test_concat.py:
```diff
@@ -219,6 +219,7 @@ def test_concat_keys_levels_no_overlap(self):
         with pytest.raises(ValueError, match=msg):
             concat([df, df2], keys=["one", "two"], levels=[["foo", "bar", "baz"]])
 
+    @pytest.mark.filterwarnings("ignore:.*append method is deprecated.*:FutureWarning")
```
remove

jreback commented on pandas/tests/reshape/concat/test_index.py:
```diff
@@ -141,6 +141,7 @@ def test_default_index(self):
         exp = DataFrame([[1, 5, 3, 7], [2, 6, 4, 8]])
         tm.assert_frame_equal(res, exp, check_index_type=True, check_column_type=True)
 
+    @pytest.mark.filterwarnings("ignore:.*append method is deprecated.*:FutureWarning")
```
remove

jreback commented on pandas/tests/reshape/merge/test_merge.py:
```diff
@@ -682,6 +682,7 @@ def _constructor(self):
 
         assert isinstance(result, NotADataFrame)
 
+    @pytest.mark.filterwarnings("ignore:.*append method is deprecated.*:FutureWarning")
```
remove

jreback reviewed (changes requested):
remove almost all of the generic filtering, the *only* ones where it should remain is the explict test_append.py

neinkeinkaffee committed:
Merge commit '4bacee5a69f0e95c19638426553eb3fa24fe6c96' into 35407-deprecate-append

neinkeinkaffee committed:
Replace append and _append by concat in tests

neinkeinkaffee committed:
Import concat from pandas instead of locally from submodule

jbrockmendel commented on pandas/tests/generic/test_duplicate_labels.py:
```diff
@@ -294,15 +294,11 @@ def test_setting_allows_duplicate_labels_raises(self, data):
         msg = "Index has duplicates."
         with pytest.raises(pd.errors.DuplicateLabelError, match=msg):
-            func(s)
+            pd.concat([pd.Series(0, index=["a", "b"]), s])
```
can you construct the series outside of the pytest.raises; makes clearer exactly what part of this line is raising

jbrockmendel commented on pandas/tests/indexing/test_partial.py:
```diff
@@ -538,7 +538,8 @@ def test_partial_set_invalid(self):
         df = orig.copy()
         df.loc["a", :] = df.iloc[0]
-        exp = orig._append(Series(df.iloc[0], name="a"))
+        s = Series(df.iloc[0], name="a")
```
"s" -> "ser" pls

jbrockmendel commented on pandas/tests/reshape/concat/test_index.py:
```diff
@@ -179,16 +178,12 @@ def test_dups_index(self):
 
         # append
-        result = df.iloc[0:8, :].append(df.iloc[8:])
+        result = concat([df.iloc[0:8, :], df.iloc[8:]])
```
are these redundant with the tests above?  this _may_ be better thought of as a targeted test for _append.

jbrockmendel commented on pandas/core/frame.py:
```diff
@@ -3267,9 +3267,12 @@ def memory_usage(self, index: bool = True, deep: bool = False) -> Series:
                 self.index.memory_usage(deep=deep), index=["Index"]
-            ).append(result)
+            )
+            result = concat([index_memory_usage, result])
```
i'd rather keep using _append internally; should be more performant and avoids circular dependency

jbrockmendel commented on pandas/tests/reshape/concat/test_append.py:
```diff
@@ -109,7 +110,8 @@ def test_append_sorts(self, sort):
-        with tm.assert_produces_warning(None):
+        # GH#35407
+        with tm.assert_produces_warning(FutureWarning):
             result = df1.append(df2, sort=sort)
```
if this file is about testing 'append', then these should just be changed to test '_append'

neinkeinkaffee committed:
Test DataFrame._append where DataFrame.append is subject under test

neinkeinkaffee committed:
Extract variable and improve variable name

neinkeinkaffee committed:
Introduce Series._append and use in test_append and frame.py

neinkeinkaffee committed:
Catch two more tests that should test _append

neinkeinkaffee committed:
Revert append -> concat and use _append internally instead

jreback commented:
@neinkeinkaffee can you merge master

neinkeinkaffee committed:
Merge master

jreback commented on doc/source/user_guide/merging.rst:
```diff
@@ -473,14 +407,20 @@ like GroupBy where the order of a categorical variable is meaningful.
-While not especially efficient (since a new object must be created), you can
-append a single row to a ``DataFrame`` by passing a ``Series`` or dict to
-``append``, which returns a new ``DataFrame`` as above.
+You can append a single row in form of a series to a ``DataFrame`` in-place
```
don't show L416, just make the alternative the thing

jreback reviewed (changes requested):
minor comment, ping on green

neinkeinkaffee committed:
Remove in-place append of single rows from the docs

jbrockmendel commented:
closes #22957?

neinkeinkaffee committed:
Merge branch 'master' into 35407-deprecate-append

neinkeinkaffee commented:
@jreback I've changed the docs and removed the part about adding rows in-place. However, the Database / Linux_py38_IO pipeline currently fails with `pytest: command not found`, while the Azure pipelines (pandas-dev.pandas Windows py38_np18_2 and Windows py39_2), fail because of a DeprecationWarning that gets raised in `pandas/tests/io/parser/test_header.py::test_no_header_prefix[pyarrow]`, which doesn't appear to be related to the changes I've made. Is this something known or should I dig into it?

jreback commented:
@neinkeinkaffee if you can merge master again; just want to get a clean run here. we do have some occasional test failures on CI (though some recent PRs do address some of these)

neinkeinkaffee committed:
Merge branch 'master' into 35407-deprecate-append

mroeschke reviewed (approved):
LGTM. Great job @neinkeinkaffee

jreback commented:
thanks @neinkeinkaffee yep very nice!
